teps/0042-taskrun-breakpoint-on-failure.md

```diff
@@ -0,0 +1,113 @@
+---
+title: taskrun-breakpoint-on-failure
+authors:
+  - "@waveywaves"
+creation-date: 2021-01-15
+last-updated: 2021-03-21
+status: proposed
+---
+
+# TEP-0042: TaskRun Breakpoint on Failure of Step 
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+<!-- /toc -->
+
+## Summary
+
+Debugging TaskRuns can be tiresome. Re-running Tasks to figure out which part of a particular step is to blame would tax productivity.
+By enabling breakpoint on failure for a TaskRun we should be able to halt a TaskRun at the failing step and get access to the
+step environment to analyze cause of the failure. Unlike legacy systems, this allows the user to debug TaskRuns while the TaskRun 
+is still running, hence improving developer productivity and Pipeline debuggability.
+
+## Motivation
+
+Lack of live debugging for Pipelines.
+
+### Goals
+
+```
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  generateName: quatredeux
+spec:
+  breakpoint:
+    onFailure: true
+```
+
+When the "breakpoint on failure" spec is mentioned in a particular TaskRun
+
+- pause TaskRun on failure of a step
+- failed step should not exit
+- user should be able to get remote shell access to the step container to analyze and debug
+
+### Non-Goals
+
+The scope of this TEP does not exceed any other kind of breakpointing or debugging apart from breakpointing on failure and failure itself.
+
+### Use Cases (optional)
+
+This enhancement can allow a more interactive way of working with TaskRuns. CLI or IDE extensions could allow users to run a 
+Task with breakpoint-on-failure and while the logs are being streamed, the user can be dropped to a shell in the breakpoint
+failed step and allow the user to debug as they develop.
+
+## Requirements
+
+* To enable breakpoint on failure of a step, it would be necessary to update the Entrypoint to support lifecycle
+changes when breakpoint on failure has been enabled.
+
+* The step environment would have to have scripts which the user can use to clear the breakpoint or
+mark the step as a success or a failure.
+
+* The entrypoint needs information about it's position in to the order of execution of steps. This will allow the
+scripts to know for sure what kind of changes need to be made to manipulate the lifecycle of the step.
+
+## Proposal
+
+TaskRuns will add a new spec called `breakpoint` under which `onFailure` can be set to true to enable breakpointing
+on failure.
+
+### Controlling step lifecycle
+
+Currently, the entrypoint writes the err postfile if there was an error during waiting for the previous postfile. This would
+be disabled so that the user who is debugging the step can control the lifecycle of a step and mark it either 
+as a success or failure by themselves.
```

Comment:
We should rephrase this a bit. In case of the *debug on failure* is set on the Taskrun, we would *hold* (for a given time) and wait for something (user connecting to the pod and issuing a command) to write to that *postfile* and continue the flow (or halt it).